File: frontend/src/core/asset-references.ts
/** Matches the public asset-reference fields in backend AssetReferences; request bodies and text remain literal. */
export const assetReferenceFields = new Set(['environmentId', 'databaseSourceId', 'apiDefinitionId', 'requirementId', 'datasetId', 'associatedCaseId', 'targetId'])

export function localAssetReferences(parentId: string | null, data: Record<string, unknown>): string[] {
  return [parentId, ...Object.entries(data).filter(([key]) => assetReferenceFields.has(key)).map(([, value]) => value)]
    .filter((value): value is string => typeof value === 'string' && value.startsWith('@'))
    .map(value => value.slice(1))
}

File: backend/src/main/java/com/aitest/asset/AssetReferences.java
package com.aitest.asset;

import java.util.Set;

/** Asset references exclude file/run identifiers and arbitrary text or request payloads. */
public final class AssetReferences {
    private AssetReferences() { }
    public static final Set<String> FIELDS = Set.of("environmentId", "databaseSourceId", "apiDefinitionId",
            "requirementId", "datasetId", "associatedCaseId", "targetId");
}
